### PR TITLE
Layout and Code Refactoring for Questions and Answers

### PR DESCRIPTION
**Pull request by filip-gregurevic**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Client: I added multiple integration tests (Jest) related to the features
- [x] Client: I added `authorities` to all new routes and check the course groups for displaying navigation elements (links, buttons)
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context

Code is hard to read with a lot of duplication.
Questions and Answers are too small for longer texts.

### Description

Refactored the code to include less methods and to be more readable.
Extracted the _question-answer_ and _question_ component. 
Changed the design of the UI.
Added the functionality to collapse the questions and answers.
Added the functionality to resize the questions and answers by dragging.
Adapted the layout on smaller screens: Questions and Answers are shown below the detail information

### Steps for Testing

1. Log in to Artemis
(1a. If not already, enable questions and answers for the course)
2. Navigate to an exercise or a lecture detail page
3. Play around with questions and answers (especially with the new drag and collapse feature)

### Screenshots

New design of Question and Answers:
Bildschirmfoto 2020-04-29 um 21 05 31

Bildschirmfoto 2020-04-29 um 21 06 55

Bildschirmfoto 2020-04-29 um 21 07 11

Bildschirmfoto 2020-04-29 um 21 07 37

Collapsed Questions and Answers:
Bildschirmfoto 2020-04-28 um 18 39 12

### CODE REVIEW

**Comment by filip-gregurevic** [comment]
@HanyaElhashemy @sascha11110 Thanks for the feedback! I implemented it now.
And another student can edit/ delete questions and answers if he is at least a tutor in that course.

File: src/main/webapp/app/overview/student-questions/student-questions.scss [hunk 1]
```diff
@@ -1,15 +1,108 @@
...
+        .card-header {
+            display: inline-flex;
+            justify-content: space-between;
+            align-items: center;
+            background-color: $background-color;
+            cursor: pointer;
+
+            .card-title {
+                display: flex;
+            }
```
Comment (sascha11110) [inline src/main/webapp/app/overview/student-questions/student-questions.scss:34]:
```suggestion
            }
            
            .row > .col-auto:last-child {
                display: flex;
                flex-direction: column;
                justify-content: center;
            }
```

File: src/main/webapp/app/overview/student-questions/student-questions.scss [hunk 2]
```diff
@@ -1,15 +1,108 @@
...
+        }
```
Comment (sascha11110) [inline src/main/webapp/app/overview/student-questions/student-questions.scss:35]:
```suggestion
        }
        
        .card-body {
            padding-top: 10px !important;
            padding-bottom: 10px !important;
            > .row >.col-12:first-child > button {
                margin: 0 15px;
            }
        }
```

File: src/main/webapp/app/overview/student-questions/student-questions.scss [hunk 3]
```diff
@@ -1,15 +1,108 @@
...
+.answer-subtitle {
     font-size: 12px;
 }
 
-.answer-approved {
-    color: green;
-    background-color: #f2f9f2;
+.question-text,
+.answer-text {
+    background: rgba(0, 0, 0, 0.1);
```
Comment (sascha11110) [inline src/main/webapp/app/overview/student-questions/student-questions.scss:66]:
```suggestion
    background: rgba(0, 0, 0, 0.1);
    padding: 8px 15px;
```

File: src/main/webapp/app/overview/student-questions/student-questions.scss [hunk 4]
```diff
@@ -1,15 +1,108 @@
...
+    border-radius: 3px;
```
Comment (sascha11110) [inline src/main/webapp/app/overview/student-questions/student-questions.scss:67]:
```suggestion
    border-radius: 3px;
    
    ::ng-deep p {
        margin: 0;
    }
```

File: src/main/webapp/app/overview/student-questions/student-questions.scss (hunk 1 above)
Comment (filip-gregurevic) [reply src/main/webapp/app/overview/student-questions/student-questions.scss:34]: I implemented that as you said in the scss

File: src/main/webapp/app/overview/student-questions/student-questions.scss (hunk 2 above)
Comment (filip-gregurevic) [reply src/main/webapp/app/overview/student-questions/student-questions.scss:41]: I implemented that in my last commit, but not using the scss but by helper classes

File: src/main/webapp/app/overview/student-questions/student-questions.scss (hunk 4 above)
Comment (filip-gregurevic) [reply src/main/webapp/app/overview/student-questions/student-questions.scss:74]: I implemented that as well in my last commit. 
And thanks for all the suggestions btw 😄

**Comment by krusche** [comment]
I don't really like the ? I would prefer if we still write Question at the top,
Same for the lightbulb, I would prefer that we write Answers

**Comment by filip-gregurevic** [comment]
> I don't really like the ? I would prefer if we still write Question at the top,
> Same for the lightbulb, I would prefer that we write Answers

@krusche I removed the icons. Now it says _question_ or _answer_

**Comment by krusche** [comment]
Can you update the screenshots in the PR? I would also suggest to move this to 4.0.3 then we have a bit more time for testing

File: src/main/webapp/app/overview/student-questions/student-question-answer/student-question-answer.component.html [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+                        <div class="col px-0">
+                            <span class="form-control-label">{{ 'artemisApp.courseOverview.exerciseDetails.faq.editAnswer' | translate }}</span>
+                        </div>
+                        <div class="col-auto px-0" *ngIf="isAtLeastTutorInCourse || isAuthorOfAnswer(studentQuestionAnswer)">
+                            <fa-icon
+                                class="mr-1 clickable"
+                                [ngbTooltip]="'artemisApp.courseOverview.exerciseDetails.faq.editAnswer' | translate"
+                                container="body"
+                                [icon]="'pencil-alt'"
+                                (click)="$event.stopPropagation(); isEditMode = false"
```
Comment (TobiasPr) [inline src/main/webapp/app/overview/student-questions/student-question-answer/student-question-answer.component.html:17]: I think angular events do not bubble like normal javascript dom events -> event.stopPropagagtion is not needed

File: src/main/webapp/app/overview/student-questions/student-question-answer/student-question-answer.component.html [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+                                *ngIf="isAtLeastTutorInCourse"
+                                [ngbTooltip]="
+                                    (studentQuestionAnswer.tutorApproved
+                                        ? 'artemisApp.courseOverview.exerciseDetails.faq.unapproveAnswer'
+                                        : 'artemisApp.courseOverview.exerciseDetails.faq.approveAnswer'
+                                    ) | translate
+                                "
+                                container="body"
+                                [icon]="'check'"
+                                (click)="$event.stopPropagation(); toggleAnswerTutorApproved()"
```
Comment (TobiasPr) [inline src/main/webapp/app/overview/student-questions/student-question-answer/student-question-answer.component.html:38]: same here

File: src/main/webapp/app/overview/student-questions/student-question-answer/student-question-answer.component.html [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+                            <fa-icon
+                                class="mr-1 clickable"
+                                [ngbTooltip]="'artemisApp.courseOverview.exerciseDetails.faq.editAnswer' | translate"
+                                container="body"
+                                [icon]="'pencil-alt'"
+                                (click)="$event.stopPropagation(); isEditMode = false"
+                            ></fa-icon>
+                            <jhi-confirm-icon
+                                class="mr-1 clickable"
+                                (click)="$event.stopPropagation()"
```
Comment (TobiasPr) [inline src/main/webapp/app/overview/student-questions/student-question-answer/student-question-answer.component.html:21]: same here

File: src/main/webapp/app/overview/student-questions/student-question-answer/student-question-answer.component.html [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+                <div class="col px-0">
+                    <span>{{ 'artemisApp.courseOverview.exerciseDetails.faq.answerLabel' | translate }}</span>
+                </div>
+                <div class="col-auto px-0" *ngIf="isAtLeastTutorInCourse || isAuthorOfAnswer(studentQuestionAnswer)">
+                    <fa-icon
+                        class="mr-1 clickable"
+                        [ngbTooltip]="'artemisApp.courseOverview.exerciseDetails.faq.editAnswer' | translate"
+                        container="body"
+                        [icon]="'pencil-alt'"
+                        (click)="$event.stopPropagation(); isEditMode = true"
```
Comment (TobiasPr) [inline src/main/webapp/app/overview/student-questions/student-question-answer/student-question-answer.component.html:74]: same here

File: src/main/webapp/app/overview/student-questions/student-question-answer/student-question-answer.component.html [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+                    ></fa-icon>
+                    <jhi-confirm-icon
+                        class="mr-1 clickable"
+                        (click)="$event.stopPropagation()"
```
Comment (TobiasPr) [inline src/main/webapp/app/overview/student-questions/student-question-answer/student-question-answer.component.html:78]: same here

File: src/main/webapp/app/overview/student-questions/student-question-answer/student-question-answer.component.html [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+                        *ngIf="isAtLeastTutorInCourse"
+                        [ngbTooltip]="
+                            (studentQuestionAnswer.tutorApproved
+                                ? 'artemisApp.courseOverview.exerciseDetails.faq.unapproveAnswer'
+                                : 'artemisApp.courseOverview.exerciseDetails.faq.approveAnswer'
+                            ) | translate
+                        "
+                        container="body"
+                        [icon]="'check'"
+                        (click)="$event.stopPropagation(); toggleAnswerTutorApproved()"
```
Comment (TobiasPr) [inline src/main/webapp/app/overview/student-questions/student-question-answer/student-question-answer.component.html:95]: same here

File: src/main/webapp/app/overview/student-questions/student-question/student-question.component.html [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+            <span>{{ 'artemisApp.courseOverview.exerciseDetails.faq.questionLabel' | translate }}</span>
+        </div>
+        <div class="col-auto px-0">
+            <fa-icon
+                class="clickable mr-1"
+                *ngIf="isQuestionAuthor || isAtLeastTutorInCourse"
+                [ngbTooltip]="'artemisApp.courseOverview.exerciseDetails.faq.editQuestion' | translate"
+                container="body"
+                [icon]="'pencil-alt'"
+                (click)="$event.stopPropagation(); isEditMode = true"
```
Comment (TobiasPr) [inline src/main/webapp/app/overview/student-questions/student-question/student-question.component.html:14]: same here

File: src/main/webapp/app/overview/student-questions/student-question/student-question.component.html [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+            ></fa-icon>
+            <jhi-confirm-icon
+                class="clickable mr-1"
+                *ngIf="isQuestionAuthor || isAtLeastTutorInCourse"
+                (click)="$event.stopPropagation()"
```
Comment (TobiasPr) [inline src/main/webapp/app/overview/student-questions/student-question/student-question.component.html:19]: same here

**Comment by filip-gregurevic** [comment]
> I tested it again and also all options in the markdown editor. Unfortunately, the coloring option does not work for me.
> 
> Apart from that, everything works perfectly!

@sascha11110 I had a look into this and it seems to be a problem with the htmlformarkdown pipe. I will fix this in another PR refactoring the whole Markdown Editor, since it isn't directly related to the questions and answers

File: src/main/webapp/app/overview/student-questions/student-question/student-question.component.html (hunk 12 above)
Comment (filip-gregurevic) [reply src/main/webapp/app/overview/student-questions/student-question/student-question.component.html]: I removed them all. I must have missed that during the refactoring.

File: src/main/webapp/app/overview/student-questions/student-question-answer/student-question-answer.component.html [hunk 13]
```diff
@@ -0,0 +1,111 @@
+<div class="row mb-1 pl-2 answer" [ngClass]="{ 'answer-approved': studentQuestionAnswer.tutorApproved }">
```
[the remark this answers is not in the document]
Comment (filip-gregurevic) [reply src/main/webapp/app/overview/student-questions/student-question-answer/student-question-answer.component.html]: I agree that it is cleaner. I now use your suggested implementation 👍

**Comment by filip-gregurevic** [comment]
> Good work :) I still have some comments: I would like to see some harder visual break between questions, because I find it hard to see where one stopes and the next one begins, but that might be just my opinion ( to make it clear: it's not something I request - it would be nice ImO but its up to you). Maybe a horizontal break, see screnshot:
> image

@TobiasPr I now added a very light background to each question (group). It should be easier now to tell them apart.

**Comment by filip-gregurevic** [comment]
> I requested changes, because I found a small bug: When I edit a question or answer and click on cancel, the change is still shown -> when I refresh it is gone, because the data did not change, but it might be confusing. Steps to reproduce:
> 
> 1. Create question
> 2. Click on edit button
> 3. Change question
> 4. Click on Cancel
> 5. Change is shown, but not applied

@TobiasPr good catch! Now the text is always reset to the initial text if not saved when closing the editor

**Comment by filip-gregurevic** [comment]
> I tested it again and also all options in the markdown editor. Unfortunately, the coloring option does not work for me.
> 
> Apart from that, everything works perfectly!

@sascha11110 I found the reason for this. I now added the 'markdown-preview' class and it should work now

**Comment by filip-gregurevic** [comment]
> Thanks for fixing the issues with the markdown editor! Looks good to me! :-)
> 
> A little suggestion (for a follow-up): would be nice if the focus is set to the corresponding textarea after clicking `Add answer` or `Add question` so that you can start typing right away.

Good idea! I'll definitely add that in a later PR

**Comment by filip-gregurevic** [comment]
> Thank you so much for addressing my previous suggestion. I still have smaller notes: could you please adapt the changes from #1381 here? The quote command does not seem to be working properly in the markdown editor.
> Also when I switch to preview tab the size of the tab shrinks, is it possible to adjust this as well?
> image
> image

@HanyaElhashemy I included the fix and the editor should always take up the whole width now

closes https://github.com/ls1intum/Artemis/issues/1057 now, too, I guess

**Pull request merged**